WalletWasabi/Tor/TorMonitor.cs

```diff
@@ -58,84 +79,145 @@ public override Task StartAsync(CancellationToken cancellationToken)
-						continue;
-					}
-					if (asyncEvent is BootstrapStatusEvent bootstrapEvent)
+					Logger.LogError($"Exception thrown when parsing event: '{reply}'", e);
+					continue;
+				}
+				if (asyncEvent is BootstrapStatusEvent bootstrapEvent)
+				{
+					BootstrapStatusEvent.Phases.TryGetValue(bootstrapEvent.Progress, out string? bootstrapInfo);
+					Logger.LogInfo($"Bootstrap progress: {bootstrapEvent.Progress}/100 ({bootstrapInfo ?? "N/A"})");
+				}
+				else if (asyncEvent is StatusEvent statusEvent)
+				{
+					if (!circuitEstablished && statusEvent.Action == StatusEvent.ActionCircuitEstablished)
 					{
-						BootstrapStatusEvent.Phases.TryGetValue(bootstrapEvent.Progress, out string? bootstrapInfo);
-						Logger.LogInfo($"Bootstrap progress: {bootstrapEvent.Progress}/100 ({bootstrapInfo ?? "N/A"})");
+						Logger.LogInfo("Tor circuit was established.");
+						circuitEstablished = true;
 					}
-					else if (asyncEvent is StatusEvent statusEvent)
+				}
+				else if (asyncEvent is CircEvent circEvent)
+				{
+					CircuitInfo info = circEvent.CircuitInfo;
+					if (!circuitEstablished && info.CircStatus is CircStatus.BUILT or CircStatus.EXTENDED or CircStatus.GUARD_WAIT)
 					{
-						if (!circuitEstablished && statusEvent.Action == StatusEvent.ActionCircuitEstablished)
-						{
-							Logger.LogInfo("Tor circuit was established.");
-							circuitEstablished = true;
-						}
+						Logger.LogInfo("Tor circuit was established.");
+						circuitEstablished = true;
 					}
-					else if (asyncEvent is CircEvent circEvent)
+					if (info.CircStatus == CircStatus.CLOSED && info.UserName is not null)
 					{
-						CircuitInfo info = circEvent.CircuitInfo;
-						if (!circuitEstablished && info.CircStatus is CircStatus.BUILT or CircStatus.EXTENDED or CircStatus.GUARD_WAIT)
-						{
-							Logger.LogInfo("Tor circuit was established.");
-							circuitEstablished = true;
-						}
-						if (info.CircStatus == CircStatus.CLOSED && info.UserName is not null)
-						{
-							Logger.LogTrace($"Tor circuit #{info.CircuitID} ('{info.UserName}') was closed.");
-							TorHttpPool.ReportCircuitClosed(info.UserName);
-						}
+						Logger.LogTrace($"Tor circuit #{info.CircuitID} ('{info.UserName}') was closed.");
+						TorHttpPool.ReportCircuitClosed(info.UserName);
 					}
 				}
 			}
-			catch (OperationCanceledException)
+		}
+		catch (OperationCanceledException)
+		{
+			if (forceTorRestartCts.IsCancellationRequested)
 			{
-				if (linkedCts.IsCancellationRequested)
+				// Attempt to shut down Tor. 
+				if (torControlClient is not null)
 				{
-					Logger.LogDebug("Tor Monitor is stopping.");
+					bool success = await ShutDownTorAsync(torControlClient).ConfigureAwait(false);
+					if (success)
+					{
+						// Wait a bit so that we don't try to connect to Tor Control until Tor is actually started.
+						await Task.Delay(2_000, CancellationToken.None).ConfigureAwait(false);
```

Comment:
Shouldn't we cancel this with `cancellationToken`? If we're shutting down WW then it should be fine to return faster.